src/tools/compiletest/src/header.rs

```diff
@@ -800,23 +789,38 @@ const DIAGNOSTICS_DIRECTIVE_NAMES: &[&str] = &[
     "unset-rustc-env",
 ];
 
-fn iter_header_extra(
+fn iter_header(
     mode: Mode,
     suite: &str,
     poisoned: &mut bool,
     testfile: &Path,
     rdr: impl Read,
-    extra_directives: &[&str],
+    // Callback arguments: (revision_name, original_line, line, line_number)
     it: &mut dyn FnMut(Option<&str>, &str, &str, usize),
```

Comment:
Maybe this should become a typedef to make it more clear as well? i.e.

```rs
/// Callback arguments: (revision_name, original_line, line, line_number)
type IterHeaderCallback = dyn FnMut(Option<&str>, &str, &str, usize);
```

then `iter_header` can become

```rs
fn iter_header(
    mode: Mode,
    suite: &str,
    poisoned: &mut bool,
    testfile: &Path,
    rdr: impl Read,
    extra_directives: &[&str],
    // Callback arguments: (revision_name, original_line, line, line_number)
    it: &mut IterHeaderCallback,
```

TBH, I still don't like how confusing this can be. Maybe introduce a struct just for the callback arguments so you don't get confused when passing args to `it`?

```rs
struct IterHeaderCallbackArgs<'rev, 'ln> {
    revision_name: Option<&'rev str>,
    /// Raw line from the test file.
    orginal_line: &'ln str,
    /// Line with the prefix up to the directive marker `//` or `//@` stripped.
    line: &'ln str,
    line_number: usize,
}
```
then each arg can be properly documented, especially to distinguish what's the difference between `original_line` and `line`.

Then when you call `it`, it can become

```rs
it(IterHeaderCallbackArgs {
    revision_name: None,
    original_line: "",
    line: directive,
    line_number: 0,
});
```

so then `iter_header` can become

```rs
fn iter_header(
    mode: Mode,
    suite: &str,
    poisoned: &mut bool,
    testfile: &Path,
    rdr: impl Read,
    extra_directives: &[&str],
    it: &mut dyn FnMut(IterHeaderCallbackArgs),
```
```

Reply:
Yeah, I like the struct idea. It's a little verbose, but it's only called in 2 places, so that's fine.

It should also make the call sites nicer, too.